design-doc.md

```diff
@@ -0,0 +1,150 @@
+
+## Problem
+
+This is an existing issue in 1-2 brownfield Azure services: Keyvault, Storage and potentially AppConfig.
+
+### Keyvault
+
+KV only has one instance in `listCertificates`, there is an optional query parameter `includePending`, which dictates whether to include pending certificates in the returned list of certificates. The service does not automatically reinject the value of `includePending` into the next link, so for all subsequent calls during paging, you will have to reinject `includePending=True` if you would like to continue seeing pending certificates in subsequent pages.
+
+### Storage
+
+Storage does a combination of continuation token paging and needing to reinject optional query parameters into the next GET calls. For example, `listBlobsInContainer` has the following paging process
+
+1. GET `<storage-container-url>/?comp=list`
+2. extract value of `NextMarker` from HTTP response body
+3. if `NextMarker` value is empty BREAK
+4. GET `<storage-container-url>/?comp=list;marker=<next-marker-value>`
+5. if `maxresults` was passed by the user, reinject this value into the GET call 
+6. GOTO 2
+
+### AppConfig
+
+I can't get in touch with anyone on the client-side team working with AppConfig since they're all no longer with us.
+
+The examples used throughout the rest of this doc will reference the `includePending` issue from KV.
+
+## Requirements
+
+1. Solves the issue for our 3 Azure services: KeyVault, Storage, AppConfig.
+2. Open question of whether this support is required for 3p. For Azure services, requiring clients to reinject parameters is against guidelines. However, should we restrict 3p authors from describing this in their specs, even if it might not be best practice?
+3. Work with templates
+
+## Design Proposals
+
+### Fixing `@nextPageOperation` in typespec-azure
+
+```tsp
+@get nextPage is Azure.Core.Foundations.Operation<{ nextLink: string, includePending?: string }, Azure.Core.Page<Certificate>>;
+
+@Azure.Core.nextPageOperation(
+    nextPage,
+    { nextLink: ResponseProperty<"nextLink">, includePending: RequestParameter<"includePending"> }
+)
+list is Azure.Core.ResourceList<Certificate>;
+```
+
+#### Pros
+
+1. Definitely easiest path forward, since all of the code is pretty much here for us. All we have to do is fix [this](https://github.com/Azure/typespec-azure/issues/1880) issue
+2. No new decorators or syntax introduced
+
+#### Cons
+
+1. Ugly definition
+2. We are gradually moving away from paging from `azure-core`, and unclear if there would be a counterpart in unbranded paging, so this won't be a good idea for 3p support or non-brownfield service support
+3. No 3p support
+
+#### Todos
+
+1. Fix [this](https://github.com/Azure/typespec-azure/issues/1880) github issue
+2. Ensure support for `@nextPageOperation` in all language emitters
+
+### Adding an incomplete next link scalar type
+
+```tsp
+scalar incompleteNextLink<ParametersToReinject[]> extends url;
```

Comment:
We should consider naming this scalar something different - `nextLinkWithAdditionalParams<>`

In the KV scenario, the next link is not *incomplete* as the next link returned by the service still works. The only issue is without `includePending` appended to the next link, the results of the next page don't include the pending certs.